pkg/chunk/composite_store.go

```diff
@@ -34,84 +34,139 @@ func (a byStart) Len() int           { return len(a) }
-		store, err := newStore(cfg, v2Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.DailyBucketsFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.DailyBucketsFrom.Time,
+			SchemaFunc: v2Schema,
+		})
 	}
 
 	if schemaCfg.Base64ValuesFrom.IsSet() {
-		store, err := newStore(cfg, v3Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.Base64ValuesFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.Base64ValuesFrom.Time,
+			SchemaFunc: v3Schema,
+		})
 	}
 
 	if schemaCfg.V4SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v4Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V4SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V4SchemaFrom.Time,
+			SchemaFunc: v4Schema,
+		})
 	}
 
 	if schemaCfg.V5SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v5Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V5SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V5SchemaFrom.Time,
+			SchemaFunc: v5Schema,
+		})
 	}
 
 	if schemaCfg.V6SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v6Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V6SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V6SchemaFrom.Time,
+			SchemaFunc: v6Schema,
+		})
 	}
 
 	if schemaCfg.V7SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v7Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V7SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V7SchemaFrom.Time,
+			SchemaFunc: v7Schema,
+		})
 	}
 
 	if schemaCfg.V8SchemaFrom.IsSet() {
-		store, err := newStore(cfg, v8Schema(schemaCfg), storage)
-		if err != nil {
-			return nil, err
-		}
-		stores = append(stores, compositeStoreEntry{schemaCfg.V8SchemaFrom.Time, store})
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V8SchemaFrom.Time,
+			SchemaFunc: v8Schema,
+		})
 	}
 
 	if schemaCfg.V9SchemaFrom.IsSet() {
-		store, err := newSeriesStore(cfg, v9Schema(schemaCfg), storage)
+		opts = append(opts, SchemaOpt{
+			From:       schemaCfg.V9SchemaFrom.Time,
+			SchemaFunc: v9Schema,
+		})
+	}
+
+	return opts
+}
+
+// StorageOpt stores when a StorageClient is to be used.
+type StorageOpt struct {
+	From   model.Time
+	Client StorageClient
+}
+
+// NewStore creates a new Store which delegates to different stores depending
+// on time.
+func NewStore(cfg StoreConfig, schemaCfg SchemaConfig, schemaOpts []SchemaOpt, storageOpts []StorageOpt) (Store, error) {
```

Comment:
Conceptually this should have very similar structure to the intervals interview question; instead of detecting overlaps you should be emitting stores.

So something like:
```go
result = []compositeStoreEntry{}
currSchema = ...
currStorage = ...

for i, j := 0; i < len(schemaOpts) && j < len(storageOpts); {
   if schemaOpts[i].From.Before(storageOpts[j].From) {
      currSchema = schemaOpts[i]
      i++
   }

   if schemaOpts[i].From.After(storageOpts[j].From) {
      currStorage = storageOpts[j]
      j++
   }

   currSchema = schemaOpts[i]
   currStorage = storageOpts[j]
   i++
   j++

   result = append(result, newStore(cfg, currSchemaOpt.SchemaFunc(schemaCfg), currStorageOpt.Client)
}

...
```

Reply:
I get what you mean, but what to advance depends on the next element and not the current element. But your idea is actually more readable and I modified the algorithm likewise.

Reply:
Looking better, but I don't see how this will produce different results to the (simpler) suggestion?  ie why does it depend on the next element?

Example:
- schema (t, v) = `[(0, 0), (10, 6), (20, 9), (40, 10)]`
- storage client (t, v) = `[(0, old), (30, new)]`
- desired output (t, schema, client) = `[(0, 0, old), (10, 6, old), (20, 9, old), (30, 9, new), (40, 10, new)]`

Perhaps extending the unit tests to include this case would also be useful.

Reply:
The more I look at this the more this looks like a merge or two sorted lists...

Reply:
Wrote this:
```
	prevSchema := schemaOpts[0]
	schemaOpts = schemaOpts[1:]
	prevStorage := storageOpts[0]
	storageOpts = storageOpts[1:]

	for len(schemaOpts) > 0 && len(storageOpts) > 0 {
		from := prevSchema.From
		if from.After(prevStorage.From) {
			from = prevStorage.From
		}

		store, err := newStore(cfg, prevSchema.SchemaFunc(schemaCfg), prevStorage.Client)
		if err != nil {
			return nil, err
		}
		stores = append(stores, compositeStoreEntry{from, store})

		// Forward the right one.
		currSchema := schemaOpts[0]
		currStorage := storageOpts[0]
		from = currSchema.From

		if from.Before(currStorage.From) {
			// Move schemaOpts by one.
			prevSchema = schemaOpts[0]
			schemaOpts = schemaOpts[1:]
		} else if from.After(currStorage.From) {
			// Move storageOpts by one.
			prevStorage = storageOpts[0]
			storageOpts = storageOpts[1:]
		}
	}

	for len(schemaOpts) == 0 && len(storageOpts) > 0 {
		store, err := newStore(cfg, prevSchema.SchemaFunc(schemaCfg), storageOpts[0].Client)
		if err != nil {
			return nil, err
		}
		stores = append(stores, compositeStoreEntry{storageOpts[0].From, store})

		storageOpts = storageOpts[1:]
	}

	for len(storageOpts) == 0 && len(schemaOpts) > 0 {
		store, err := newStore(cfg, schemaOpts[0].SchemaFunc(schemaCfg), prevStorage.Client)
		if err != nil {
			return nil, err
		}
		stores = append(stores, compositeStoreEntry{schemaOpts[0].From, store})

		schemaOpts = schemaOpts[1:]
	}
```

While it seems to work, `(30, 9, new)` doesn't come in. And there needs to be an extra check. At this point, I think I like the current code better.

Reply:
So I added the test. Should I push the code where I track `prev*` instead of `next*`?